ffi_internals/src/type_ffi.rs

```diff
@@ -0,0 +1,530 @@
+//!
+//! Contains structures describing the data for a Rust type, and implementations for building the
+//! related FFI and consumer implementations.
+//!
+
+use crate::{
+    parsing,
+    parsing::{FieldAttributes, WrappingType},
+};
+use heck::SnakeCase;
+use proc_macro2::TokenStream;
+use proc_macro_error::{abort, OptionExt};
+use quote::{format_ident, quote};
+use syn::{spanned::Spanned, Ident, Type};
+
+static STRING: &str = "String";
+static STR: &str = "str";
+static DATETIME: &str = "NaiveDateTime";
+static UUID: &str = "Uuid";
+static BOOL: &str = "bool";
+static U8: &str = "u8";
+static U16: &str = "u16";
+static U32: &str = "u32";
+static U64: &str = "u64";
+static I8: &str = "i8";
+static I16: &str = "i16";
+static I32: &str = "i32";
+static I64: &str = "i64";
+static F32: &str = "f32";
+static F64: &str = "f64";
+
+/// Describes a Rust type that is exposed via FFI (as the type of a field, or the type returned by a
+/// function, or a function parameter, etc).
+///
+#[derive(Debug, Clone, PartialEq)]
+pub enum TypeIdentifier {
+    /// A type that should be exposed behind on opaque pointer; we'll make this available as a
+    /// `*const T`, and consumers of that interface will be able to initialize, free, and access
+    /// properties on the type from getter functions.
+    ///
+    Boxed(Ident),
+    /// A timestamp that's safe to expose across the FFI (see `ffi_core::datetime`).
+    ///
+    DateTime,
+    /// A type that should be exposed as a raw value (like an i32, or a repr(C) enum).
+    ///
+    Raw(Ident),
+    /// A String.
+    ///
+    String,
+    /// A Uuid.
+    ///
+    Uuid,
+}
+
+impl From<Ident> for TypeIdentifier {
+    fn from(type_path: Ident) -> Self {
+        match type_path {
+            t if t == DATETIME => Self::DateTime,
+            t if t == STRING || t == STR => Self::String,
+            t if t == UUID => Self::Uuid,
+            t if t == BOOL
+                || t == U8
+                || t == U16
+                || t == U32
+                || t == U64
+                || t == I8
+                || t == I16
+                || t == I32
+                || t == I64
+                || t == F32
+                || t == F64 =>
```

Comment:
These `if`s feel kinda weird here... what about something along the lines of

```rust
match &type_path.to_string() {
    DATETIME => Self::DateTime,
    STRING | STR => Self::String,
    // etc
}
```

Reply:
Yep that's much less weird, just had to switch those from `static` to `const`.